sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/PartitionMetadataCache.scala

```diff
@@ -0,0 +1,304 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+package com.azure.cosmos.spark
+
+import com.azure.cosmos.implementation.{CosmosClientMetadataCachesSnapshot, Strings}
+import com.azure.cosmos.models.{CosmosChangeFeedRequestOptions, FeedRange}
+import com.azure.cosmos.spark.CosmosPredicates.{assertNotNull, assertNotNullOrEmpty, assertOnSparkDriver, requireNotNull}
+import com.fasterxml.jackson.databind.node.ObjectNode
+import org.apache.spark.broadcast.Broadcast
+import reactor.core.scala.publisher.SMono
+import reactor.core.scala.publisher.SMono.PimpJMono
+import reactor.core.scheduler.Schedulers
+
+import java.time.Instant
+import java.util.{Timer, TimerTask}
+import java.util.concurrent.atomic.{AtomicLong, AtomicReference}
+import scala.collection.concurrent.TrieMap
+
+// The partition metadata here is used purely for a best effort
+// estimation of number of Spark partitions needed for a certain
+// physical partition - it is not guaranteeing functional correctness
+// because the cached metadata could be old or even be for a different
+// container after deletion and recreation of a container
+private object PartitionMetadataCache extends CosmosLoggingTrait {
+  private[this] val Nothing = 0
+  private[this] val cache = new TrieMap[String, PartitionMetadata]
+  private[this] var cacheTestOverride: Option[TrieMap[String, PartitionMetadata]] = None
+
+  // purpose of the time is to update partition metadata
+  // additional throughput when more RUs are getting provisioned
+  private[this] val timerName = "partition-metadata-refresh-timer"
+  private[this] val timerOverrideName = "partition-metadata-refresh-timerOverride"
+  private[this] val timer: Timer = new Timer(timerName, true)
```

Comment:
Timer is a very old concurrency api in java 5.
any reason not to use `ScheduledThreadExecutor` or `ExecutorService`

using `ScheduledThreadPoolExecutor` is usually preferred as you would have better control over the threading, etc.
https://stackoverflow.com/questions/409932/java-timer-vs-executorservice

Reply:
Will look into this in separate PR - added a TODO comment for tracking for now. I know for sure that the Timer configured here is behaving as expected (used the same model in current OLTP connector) - but can look into optimizing it later.

Reply:
But to answer your question - no just unawareness. Used the concept that had worked reasonably well in current Connector version :-) I had one issue there when I did not make sure to use a daemon thread - but since that has been fixed it worked well - and the logging made clear which activity was happening on the timer thread. It also allowed me to avoid reentrance - so had everything I needed.